Extend range window queries to non-timestamp order-by columns

mythrocks opened the pull request:
(Redux of #7675, which was for `branch-0.19`, and included support to scale duration scalars to the orderby column's resolution.)

This change introduces a `grouped_range_rolling_window()` function, to generalize the functionality available through `grouped_time_range_rolling_window()`.

### Prelude
Recall that `grouped_time_range_rolling_window()` applies an aggregation over a dynamic window of rows. The width of the window is determined based on a timestamp column, and the relative preceding/following offset specified in days. E.g. Consider this dataset:

```
[ // user,  sales_amt,  YYYYMMDD (date)
  { "user1",   10,      20200101    },
  { "user2",   20,      20200101    },
  { "user1",   20,      20200102    },
  { "user1",   10,      20200103    },
  { "user2",   30,      20200101    },
  { "user2",   80,      20200102    },
  { "user1",   50,      20200107    },
  { "user1",   60,      20200107    },
  { "user2",   40,      20200104    }
]
```
Grouping by `user_id`, and ordering by `date` yields the following `sales_amt` vector (with 2 groups, one for each distinct `user_id`):
```
Date :(202001-)  [ 01,  02,  03,  07,  07,    01,   01,   02,  04 ]
Input:           [ 10,  20,  10,  50,  60,    20,   30,   80,  40 ]
                   <-------user1-------->|<---------user2--------->
```
The `SUM` aggregation applied over a window of `(1 DAY PRECEDING, 1 DAY FOLLOWING)` with a minimum of 1 period yields the following result column:
```
Results:         [ 30,  40,  30,  110, 110,  130,  130,  130,  40 ]
```
### What's new in this commit
`grouped_range_rolling_window()` extends the erstwhile range-window functionality in two ways:
  1. The order-by column (`date`) is **not restricted to timestamps**. This may also be a sorted integral column. The preceding/following offsets will then need to be specified as scalars of the same type as the order-by column.
  E.g. 
     ```
     [ // user,  sales_amt,  seconds_since_snap
       { "user1",   10,         86451          },
       { "user2",   20,         86735          },
       { "user1",   20,         89162          },
       { "user1",   10,         92152          },
       ...]
     ```
  2. If the order-by column is indeed a timestamp, the **preceding/following offsets may be duration scalars**, whose precision must match that of the timestamp order-by column.
  E.g. If `orderby_column.type() == TIMESTAMP_SECONDS`, `preceding.type()` may only be `DURATION_SECONDS`.

Analogous to `window_bounds`, a new `range_window_bounds` class has been introduced to specify bounds for range window functions: 
1. Supports scalar values for preceding/following 
2. Supports `UNBOUNDED` window widths
   
`range_window_bounds` currently supports only integral and duration scalars.
Correspondingly, the `orderby_column` can only be integral and timestamp columns. Support for floats and fixed_point types may be added at a later date.

The existing `grouped_time_range_rolling_window()` function now delegates to the same backend as `grouped_range_rolling_window()`.

mythrocks commented:
Repeating [comments from #7675](https://github.com/rapidsai/cudf/pull/7675#issuecomment-805088357) for clarity:

> we should also consider if there is redundancy in the existing APIs. E.g., `grouped_rolling_window` has overloads that take both `size_type` and `window_bounds` for the preceding/following window sizes. Isn't `window_bounds` more generic and therefore could replace the `size_type` overload?

I agree with @jrhemstad's assessment above. I hope to replace `grouped_time_range_rolling_window()` completely with `grouped_range_rolling_window()`. 

I envision the following changes in an upcoming release:
1. Deprecate (and later, remove) both `grouped_time_range_rolling_window()` functions. These are an artifact of the solving smaller parts of the more generic problem. `grouped_range_rolling_window()` will subsume this feature.
2. Deprecate (and later, remove) the `size_type` overloads of `grouped_rolling_window()`. These do not support `UNBOUNDED` ranges. We should use `window_bounds` for fixed width window functions.
3. Deprecate (and later, remove) `rolling_window()` and `grouped_rolling_window()` overloads that do not take `default_values`. These exist currently because they were added before `LEAD`/`LAG` were supported.
4. Possibly introduce `window_bounds` overloads for `rolling_window()`, if the Python/Pandas crew do not object, and deprecate the older functions. Without this, `UNBOUNDED` windows can't be supported with *most* `rolling_window()` functions.

This work hasn't been tackled yet because I'm still exploring what else we might need to support more non-trivial ranking functions. E.g. `RANK()`/`DENSE_RANK()` requires access to columns that the current interfaces do not have access to.

wbo4958 commented:
FYI, The java bindings PR https://github.com/rapidsai/cudf/pull/7909

mike-wendt reviewed (approved):
Approved changes for ops `CODEOWNERS` files; will rely on team to verify other changes

nvdbaranec commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -379,6 +381,125 @@ std::unique_ptr<column> grouped_time_range_rolling_window(
+ * @param[in] aggr The rolling window aggregation type (SUM, MAX, MIN, etc.)
+ *
+ * @returns   A nullable output column containing the rolling window results
+ */
```
Now this is documentation 👍

mythrocks replied:
:D

Oh, you're being serious?

nvdbaranec commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -715,115 +720,161 @@ std::unique_ptr<column> time_range_window_DESC(
+    CUDF_FAIL("Expected timestamp orderby column.");
+  }
+};
+
+range_window_bounds to_range_bounds(cudf::size_type num_days, cudf::data_type timestamp_type)
```
Docs

mythrocks replied:
Point taken.
(I was trying to be sneaky about not documenting short utility/implementation functions in the `unnamed` namespace.)

nvdbaranec commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -715,115 +720,161 @@ std::unique_ptr<column> time_range_window_DESC(
+{
+  return cudf::type_dispatcher(timestamp_type, to_duration_bounds{}, num_days);
+}
+
+range_window_bounds to_range_bounds(cudf::window_bounds const& days_bounds,
```
Docs

nvdbaranec commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -715,115 +720,161 @@ std::unique_ptr<column> time_range_window_DESC(
-                                                          std::unique_ptr<aggregation> const& aggr,
-                                                          rmm::cuda_stream_view stream,
-                                                          rmm::mr::device_memory_resource* mr)
+std::unique_ptr<column> grouped_range_rolling_window(table_view const& group_keys,
```
Copydoc from the header.

nvdbaranec commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -878,16 +922,18 @@ std::unique_ptr<column> grouped_time_range_rolling_window(table_view const& grou
                                                           std::unique_ptr<aggregation> const& aggr,
```
Copydoc from header.

nvdbaranec commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -379,6 +381,125 @@ std::unique_ptr<column> grouped_time_range_rolling_window(
   std::unique_ptr<aggregation> const& aggr,
```
Not part of this PR, but this copydoc block seems incorrect.  `preceding_window_in_days` and `following_window_in_days` are a different type.

mythrocks replied:
I thought this might be acceptable, since the description still fits. All that changes between the two is that the second supports `UNBOUNDED` window bounds.
I was trying to avoid having to repeat the detailed description for each function. I can try redoing this.

mythrocks replied:
I've switched to using `@copydetails`, and redone the parameter descriptions.

nvdbaranec commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -878,16 +922,18 @@ std::unique_ptr<column> grouped_time_range_rolling_window(table_view const& grou
+                                      mr);
 }
 
 std::unique_ptr<column> grouped_time_range_rolling_window(table_view const& group_keys,
```
Copydoc.

nvdbaranec commented on cpp/src/rolling/range_window_bounds.cpp:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2021, NVIDIA CORPORATION.
```

nvdbaranec commented on cpp/include/cudf/rolling/range_window_bounds.hpp:
```diff
@@ -0,0 +1,83 @@
+    assert_invariants();
+  }
+
+  void assert_invariants() const;
```
The code in here only ever seems to be used in the above private constructor.  Does it need to be a separate function?

mythrocks replied:
> Does it need to be a separate function?

No, sir. Not any more. I'll collapse this back into the constructor.

This is vestigial, from a time when there were non-const methods in `range_window_bounds` for scaling the bounds to a different/lower precision. Those used to `assert_invariants()` at the end.

mythrocks replied:
Something is wrong here. I thought I removed `assert_invariants()`. Unresolving, until I have double checked.

harrism commented:
@gpucibot merge

vuule commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -205,57 +207,62 @@ std::unique_ptr<column> grouped_rolling_window(table_view const& group_keys,
-/// multiplied by `24*60*60`, before comparisons with the timestamps.
-size_t multiplication_factor(cudf::data_type const& data_type)
+template <typename T, std::enable_if_t<!std::numeric_limits<T>::is_signed>* = nullptr>
+__device__ T add_and_check_overflow(T const& value, T const& delta)
```
Maybe `add_safe`? I've seen similarly named functions in the code base, maybe the meaning more more specific than I understood.

vuule commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -205,57 +207,62 @@ std::unique_ptr<column> grouped_rolling_window(table_view const& group_keys,
+}
+
+template <typename T, std::enable_if_t<std::numeric_limits<T>::is_signed>* = nullptr>
+__device__ T subtract_and_check_underflow(T const& value, T const& delta)
```
similar here

vuule commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -273,28 +280,29 @@ std::unique_ptr<column> expand_to_column(Calculator const& calc,
 {
   size_type nulls_begin_idx, nulls_end_idx;
-  std::tie(nulls_begin_idx, nulls_end_idx) = get_null_bounds_for_timestamp_column(timestamp_column);
+  std::tie(nulls_begin_idx, nulls_end_idx) = get_null_bounds_for_orderby_column(orderby_column);
```
No need to use `std::tie` now :)
```suggestion
  auto [nulls_begin_idx, nulls_end_idx] = get_null_bounds_for_orderby_column(orderby_column);
```

mythrocks replied:
(Repeating inline for context.)

I appreciate the suggestion to use structured bindings, but I'm afraid they don't suffice in this context: I would have to use those as captures for the lambdas that follow, and that is unfortunately disallowed.
```
../src/rolling/grouped_rolling.cu(359): error: structured binding cannot be captured
          detected during:
            instantiation of "std::unique_ptr<cudf::column, std::default_delete<cudf::column>> cudf::<unnamed>::grouped_r
ange_rolling_window_impl<OrderByT>(const cudf::column_view &, const cudf::column_view &, const cudf::order &, const rmm::
device_uvector<cudf::size_type> &, const rmm::device_uvector<cudf::size_type> &, cudf::range_window_bounds, cudf::range_w
indow_bounds, cudf::size_type, const std::unique_ptr<cudf::aggregation, std::default_delete<cudf::aggregation>> &, rmm::c
uda_stream_view, rmm::mr::device_memory_resource *) [with OrderByT=int8_t]"
```

vuule commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -384,23 +390,23 @@ get_null_bounds_for_timestamp_column(column_view const& timestamp_column,
 
     // Null timestamps exist. Find null bounds, per group.
     thrust::for_each(
       thrust::device,
```
Not using the stream here
```suggestion
      rmm::exec_policy(stream),
```

vuule commented on cpp/src/rolling/grouped_rolling.cu:
```diff
@@ -715,115 +720,198 @@ std::unique_ptr<column> time_range_window_DESC(
+{
+  return days_bounds.is_unbounded
+           ? range_window_bounds::unbounded(
+               data_type{static_cast<cudf::type_id>(static_cast<int32_t>(timestamp_type.id()) + 5)})
```
What does the 5 mean here?

vuule commented on cpp/src/rolling/range_window_bounds.cpp:
```diff
@@ -0,0 +1,70 @@
+
+struct range_scalar_constructor {
+  template <typename T, std::enable_if_t<!detail::is_supported_range_type<T>(), void>* = nullptr>
+  scalar* operator()(scalar const& range_scalar_) const
```
It's preferred to have factory methods return unique_ptr. This makes the ownership transfer explicit. Similarly, range_window_bounds should take a unique_ptr since it's a sink for the scalar.

mythrocks replied:
>  (it's a bit scary to see a new call, especially without any delete calls).

Hmm. I do see your point about this being not being transparent. I'll check if I can restructure this a little better, and try do away with the raw pointer.

mythrocks replied:
I have switched the factory implementation to return `unique_ptr`, which I then `std::move()` to a `shared_ptr`. Thank you for this suggestion.

vuule commented on cpp/include/cudf/rolling/range_window_bounds.hpp:
```diff
@@ -0,0 +1,77 @@
+
+ private:
+  const bool _is_unbounded{false};
+  std::shared_ptr<scalar> _range_scalar{nullptr};
```
Does this need to be a `shared_ptr`? I don't see where the ownership is shared.

vuule commented on cpp/src/rolling/range_window_bounds_detail.hpp:
```diff
@@ -0,0 +1,139 @@
+template <typename T, std::enable_if_t<!std::numeric_limits<T>::is_signed, void>* = nullptr>
+void assert_non_negative(T const& value)
+{
+  // Unsigned values are non-negative.
```
Is this done because we would get a compiler warning if we checked the sing of an unsigned variable?

mythrocks replied:
Yes, sir. That's the case.

vuule replied:
You should be able to replace the SFINAE with constexpr if, implement the assert funciton this way:
```
    if constexpr (std::is_signed<T>::value)
        CUDF_EXPECTS(value >= T{0}, "Range scalar must be >= 0.");
```

mythrocks replied:
Yay, C++17! Let me try that.

mythrocks replied:
Thanks for this suggestion, @vuule. (The switch to C++17 has been brilliant thus far.)

vuule reviewed (changes requested):
Impressive docs and tests :)
Got some very minor suggestions.

mythrocks commented:
@vuule, thank you for the detailed review. GitHub is refusing to let me reply inline, so I'm going to have to quote and reply here, manually. Please bear with me:

> It's usually better to avoid negatives in the names... `!is_unbounded` adds a bit of overhead when reading the code.

I heartily agree with this advice, in general contexts. I'm afraid the term `unbounded` comes from [SQL syntax for `OVER` clauses in window functions](https://stevestedman.com/2012/03/transact-sql-over-clause-rows-unbounded-preceding-or-following/), indicating that the bounds go all the way to the beginning/end of the group/column.
The problem with un-negating `unbounded` is that it amounts to double-negation for those familiar with window functions. :] Could I please request permission to leave this, err, unchanged?

> Should the default be `unbounded`, given that the default range is `nullptr`?

That's a good catch. I'll change this.

> why is the helper needed?

Hmm... This *was* so that I didn't have to explicitly state the template arguments for `window_exec_impl`. It was for the same reason that `std::back_inserter()` is a type-deducing helper for `std::back_insert_iterator<>`. But since we're now on C++17, I might be able to do without the helper. I'll check.

mythrocks commented:
> Maybe `add_safe`?

This function (and the other one) aims to prevent overflow/underflow in bounds calculation. I caught this when I enabled range queries on unsigned `order_by` columns, and got incorrect results. The aim is to cap the min-value at `numeric_limits<>::min()` and  max-value at `numeric_limits<>::max()`.
I have changed these to `add_safe()` and `subtract_safe()`.

> No need to use `std::tie` now :)

I appreciate the suggestion to use structured bindings, but I'm afraid they don't suffice in this context: I would have to use those as captures for the lambdas that follow, and that is unfortunately disallowed.
```
../src/rolling/grouped_rolling.cu(359): error: structured binding cannot be captured
          detected during:
            instantiation of "std::unique_ptr<cudf::column, std::default_delete<cudf::column>> cudf::<unnamed>::grouped_r
ange_rolling_window_impl<OrderByT>(const cudf::column_view &, const cudf::column_view &, const cudf::order &, const rmm::
device_uvector<cudf::size_type> &, const rmm::device_uvector<cudf::size_type> &, cudf::range_window_bounds, cudf::range_w
indow_bounds, cudf::size_type, const std::unique_ptr<cudf::aggregation, std::default_delete<cudf::aggregation>> &, rmm::c
uda_stream_view, rmm::mr::device_memory_resource *) [with OrderByT=int8_t]"
```

> Not using the stream here

Thanks for pointing that out. I keep forgetting to correct that. Fixed now.

> What does the 5 mean here?

You will hate me for this: this was shorthand to convert from `TIMESTAMP_DAYS` to `DURATION_DAYS`, `TIMESTAMP_SECONDS` to `DURATION_SECONDS` etc., without using a switch-case. :/ I'll make this more explicit.

vuule commented:
I wonder why you cannot reply inline, I can reply to myself.
Feel free to keep `is_unbounded`, I was not aware of the domain terminology.

mythrocks commented:
> It's preferred to have factory methods return unique_ptr...
> range_window_bounds should take a unique_ptr
> Does this need to be a `shared_ptr`? 

Please note that `range_window_bounds` has a private constructor, and is supposed to be constructed by one of two factory methods:
1. `get(scalar const&)`
2. `unbounded(data_type)`

The problem was that we'd need to make a copy of the scalar to store in `range_window_bounds`, and to make `range_window_bounds` itself copyable. The `shared_ptr` allows this cleanly.

> It's preferred to have factory methods return unique_ptr...
> range_window_bounds should take a unique_ptr

Since these functions are both private to implementation, and used only to initialize the `shared_ptr`, could I please keep the `scalar*`?

vuule commented:
> Since these functions are both private to implementation, and used only to initialize the shared_ptr, could I please keep the scalar*?

I guess it's fine since you already have 3 approvals :) Please add comments to explain that the ownership is transferred in the constructor, and that the allocated object will be freed via the shared pointer (it's a bit scary to see a `new` call, especially without any `delete` calls).

mythrocks commented on cpp/include/cudf/rolling/range_window_bounds.hpp:
```diff
@@ -0,0 +1,77 @@
+   * @return true If window is unbounded
+   * @return false If window is of finite bounds
+   */
+  bool is_unbounded() const { return _is_unbounded; }
```
> Feel free to keep is_unbounded, I was not aware of the domain terminology.

Thanks, @vuule. I'll resolve this now.

mythrocks commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -379,6 +401,125 @@ std::unique_ptr<column> grouped_time_range_rolling_window(
+ * Currently, only the following combinations of `orderby` column type and range types
+ * are supported:
+ *   1. If `orderby` column is a TIMESTAMP, the `preceding`/`following` windows are specified
+ *      in terms of lower resolution `DURATION` scalars.
```
This is wrong: I thought this had been fixed already.

In the current implementation, we no longer support duration-types with lower resolution than the orderby column. The resolutions must be identical. `DURATION_DAYS` for `TIMESTAMP_DAYS`, `DURATION_SECONDS` for `TIMESTAMP_SECONDS`, `INT32` for `INT32`.

I will correct this shortly.

mythrocks replied:
Done.

mythrocks commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -379,6 +401,125 @@ std::unique_ptr<column> grouped_time_range_rolling_window(
+ * For the following input:
+ *
+ *  [ // driver_name,  num_overtakes,  lap_number
+ *    {   "bottas",        1,            1        },
```
Nitpick: In the interest of uniformity, the example should choose either the first-name or the last-name in the dataset. Either change `bottas` to `valtteri`, or change `lewis` to `hamilton`.

mythrocks replied:
Done. `lewis` -> `hamilton`.

mythrocks commented:
> (it's a bit scary to see a `new` call, especially without any `delete` calls).

@vuule, I have changed the (private) factory implementation to return `shared_ptr<>` instead of naked pointers, and tried to clarify this in documentation. I'm hoping this is more palatable.

vuule commented:
> > (it's a bit scary to see a `new` call, especially without any `delete` calls).
> 
> @vuule, I have changed the (private) factory implementation to return `shared_ptr<>` instead of naked pointers, and tried to clarify this in documentation. I'm hoping this is more palatable.

That's better IMO :) Any specific reason why you used shared_ptr instead of unique_ptr?

mythrocks commented:
> That's better IMO :) Any specific reason why you used `shared_ptr` instead of `unique_ptr`?

The initial version of this PR (#7675) used `unique_ptr`. The hard part was how the caller would specify the scalar (which is polymorphic), and the lifetime/ownership of said scalar. Using `unique_ptr` disables copyability of `range_window_bounds`, and precludes the use of the same `range_window_bounds` object for multiple window function calls.
(#7675 included an abortive attempt to use `range_window_bounds&&` in the API, and too many `std::move()` calls.)

Adopting a scalar clone via `shared_ptr` turns out to be cleaner, overall. (I doff my hat to @nvdbaranec for the suggestion.)

vuule reviewed (approved):
Thank you for addressing the comments!

mythrocks commented:
@gpucibot merge

mythrocks commented:
> @gpucibot merge

Let me try that again. _**Ahem**_...

mythrocks commented:
Looks like we're waiting for a `+1` from `rapidsai/ops-codeowners`.
